google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/functions/WindowsPasswordGenerator.java

```diff
@@ -0,0 +1,211 @@
+
+import javax.annotation.Resource;
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.jclouds.compute.reference.ComputeServiceConstants;
+import org.jclouds.crypto.Crypto;
+import org.jclouds.googlecomputeengine.GoogleComputeEngineApi;
+import org.jclouds.googlecomputeengine.domain.Instance;
+import org.jclouds.googlecomputeengine.domain.Instance.SerialPortOutput;
+import org.jclouds.googlecomputeengine.domain.Metadata;
+import org.jclouds.googlecomputeengine.domain.Operation;
+import org.jclouds.googlecomputeengine.features.InstanceApi;
+import org.jclouds.logging.Logger;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.base.Splitter;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.io.BaseEncoding;
+import com.google.common.util.concurrent.Atomics;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.google.inject.TypeLiteral;
+
+/**
+ * References:
+ * <ul>
+ *   <li>{@linkplain https://cloud.google.com/compute/docs/instances/automate-pw-generation}
+ *   <li>{@linkplain https://github.com/GoogleCloudPlatform/compute-image-windows/blob/master/examples/windows_auth_java_sample.java}
+ * </ul>
+ * 
+ * In brief, the sequence is:
+ * <ol>
+ *   <li>Generate a temporary key for encrypting and decrypting the password
+ *   <li>Send the RSA public key to the instance, by settings its metadata
+ *   <li>Retrieve the result from the {@link SerialPortOutput}
+ *   <li>Decode and decrypt the result.
+ * </ol>
+ */
+public class WindowsPasswordGenerator implements Function<AtomicReference<Instance>, String> {
+
+   /**
+    * Indicates when the key should expire. Keys are one-time use, so the metadata doesn't need to stay around for long.
+    * 5 minutes chosen to allow for differences between time on the client
+    * and time on the server.
+    */
+   private static final long EXPIRE_DURATION = 30 * 60 * 1000;
+
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   protected Logger logger = Logger.NULL;
+
+   private final GoogleComputeEngineApi api;
+   private final Crypto crypto;
+   private final Predicate<AtomicReference<Operation>> operationDone;
+   
+   @Inject
+   protected WindowsPasswordGenerator(GoogleComputeEngineApi api, Crypto crypto, Predicate<AtomicReference<Operation>> operationDone) {
+      this.api = api;
+      this.crypto = crypto;
+      this.operationDone = operationDone;
+   }
+
+   @Override
+   public String apply(AtomicReference<Instance> instance) {
+      // FIXME get zone how?
+      String zone;
+      //zoneUri = instance.get().zone();
+      zone = "us-central1-a";
```

Comment:
I think we should reuse something like [zoneUri](https://github.com/andreaturli/jclouds-labs-google/blob/f06fb58ccd08d64c9b2fc66c5915df647fabb4ba/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/GoogleComputeEngineServiceAdapter.java#L288-L288)